src/vehicle_gui.cpp

```diff
@@ -110,30 +166,70 @@ uint GetUnitNumberDigits(VehicleList &vehicles)
 		unitnumber = max<uint>(unitnumber, (*v)->unitnumber);
 	}
 
-	if (unitnumber >= 10000) return 5;
-	if (unitnumber >=  1000) return 4;
-	if (unitnumber >=   100) return 3;
-
-	/*
-	 * When the smallest unit number is less than 10, it is
-	 * quite likely that it will expand to become more than
-	 * 10 quite soon.
-	 */
-	return 2;
+	return CountDigitsForAllocatingSpace(unitnumber);
 }
 
 void BaseVehicleListWindow::BuildVehicleList()
 {
-	if (!this->vehicles.NeedRebuild()) return;
+	if (!this->vehgroups.NeedRebuild()) return;
 
 	DEBUG(misc, 3, "Building vehicle list type %d for company %d given index %d", this->vli.type, this->vli.company, this->vli.index);
 
+	/* Note: This can be optimized so that merely changing the "group by" dropdown value does not need to call GenerateVehicleSortList again. */
 	GenerateVehicleSortList(&this->vehicles, this->vli);
 
-	this->unitnumber_digits = GetUnitNumberDigits(this->vehicles);
+	this->vehgroups.Clear();
+	if (this->grouping == GB_NONE) {
+		uint max_unitnumber = 0;
+		for (auto it = this->vehicles.Begin(); it != this->vehicles.End(); ++it) {
+			GUIVehicleGroup * const vehgroup = this->vehgroups.Append();
+
+			vehgroup->vehicles_begin = it;
+			vehgroup->vehicles_end = it + 1;
+			vehgroup->display_profit_this_year = (*it)->GetDisplayProfitThisYear();
+			vehgroup->display_profit_last_year = (*it)->GetDisplayProfitLastYear();
+			vehgroup->age = (*it)->age;
+
+			max_unitnumber = max<uint>(max_unitnumber, (*it)->unitnumber);
+		}
+		this->unitnumber_digits = CountDigitsForAllocatingSpace(max_unitnumber);
+	} else {
+		std::stable_sort(this->vehicles.Begin(), this->vehicles.End(), [](const Vehicle * const &u, const Vehicle * const &v) {
+			return u->FirstShared() < v->FirstShared(); // Sort by the primary vehicle; we just want all vehicles that share the same orders to form a contiguous range.
+		});
+
+		uint max_num_vehicles = 0;
+
+		const Vehicle **begin = this->vehicles.Begin();
+		while (begin != this->vehicles.End()) {
+			const Vehicle **end = std::find_if_not(begin, this->vehicles.End(), [first_shared = (*begin)->FirstShared()](const Vehicle * const &v) {
+				return v->FirstShared() == first_shared;
+			});
+
+			GUIVehicleGroup * const vehgroup = this->vehgroups.Append();
+
+			vehgroup->vehicles_begin = begin;
+			vehgroup->vehicles_end = end;
+			vehgroup->display_profit_this_year = std::accumulate(begin, end, static_cast<Money>(0), [](const Money &sum, const Vehicle* const &v) {
+				return sum + v->GetDisplayProfitThisYear();
+			});
+			vehgroup->display_profit_last_year = std::accumulate(begin, end, static_cast<Money>(0), [](const Money &sum, const Vehicle* const &v) {
+				return sum + v->GetDisplayProfitLastYear();
+			});
+			vehgroup->age = std::accumulate(begin, end, static_cast<Date>(0), [](const Date &largest, const Vehicle* const &v) {
```

Comment:
As much as I like the std::accumulate, perhaps a single loop would be nicer here, instead of 3?